### DISCUSSION TITLE
What is the difference between the buffer_ and component_ metrics/telemetry?

### DISCUSSION THREAD

**Discussion by alex-rowe** [vectordotdev/vector#18615] [Q&A] [answered]
We are looking to monitor our sinks for any potential errors but I can't work out the exact difference between the `buffer_` and `component_` telemetry metrics.

Do the `buffer` metrics roll up into the `component` metrics?

For example, if we have a buffer configured and set to `drop_newest`, what metric does that show under? does it show under `buffer_discarded_events_total` or `component_discarded_events_total` or both? For this example, we'd want to alert on any events being dropped by the sink.

Similarly, if we wanted to monitor a kind of throughput, would we use `buffer_sent_events_total` or `component_sent_events_total`? For this example, we'd want to alert on something like "is sent volume over X less than the average volume"

**Comment by jszwedko** [comment, marked as answer]
Hi @alex-rowe !

The `buffer_*` metrics only apply to the buffers and are not rolled up into the `component_*` metrics. In your case, if the buffer is the one dropping events, it will appear in `buffer_discarded_events_total`.

**Reply by alex-rowe** [reply]
Thanks for the update @jszwedko 

What are the scenarios for the `buffer` or `component` dropping events?

For `buffer`, is that just once it's reached it's max size/events value?

Then for the `sent_events` style metrics. Is the flow, 10 events buffer received -> 10 buffer sent -> 10 component received -> 10 component sent? Which would be the final metric to check for "this sink sent this amount of events successfully"?

**Reply by jszwedko** [reply]
> For buffer, is that just once it's reached it's max size/events value?

Correct, that should only happen if `buffer.on_full = "drop_newest"` and the buffer is overflowing.

For `component_discarded_events` it depends on the component. If the `intentional` metric label is `false` this generally indicates an error occurred. The `filter` transform, for example, increments that counter when filtering events (with `intentional` set to `true`).

> Then for the `sent_events` style metrics. Is the flow, 10 events buffer received -> 10 buffer sent -> 10 component received -> 10 component sent? Which would be the final metric to check for "this sink sent this amount of events successfully"?

For this, `component_sent_events_total` would be what you want.

**Reply by alex-rowe** [reply]
The component we were looking at was `aws_kinesis_streams`. I was probably looking in the wrong place, but I couldn't see where the `intentional` metric label was configured for this one.

**Reply by jszwedko** [reply]
It _should_ have an `intentional` label on `component_discarded_events_total`.

**Reply by alex-rowe** [reply]
Is this where they're defined [src/internal_events/aws_kinesis.rs#L32](https://github.com/vectordotdev/vector/blob/a6b1bed163fa52e1a3fc2189ac24f0f355d752cf/src/internal_events/aws_kinesis.rs#L32)?

I couldn't see anything on the docs for the sink [aws_kinesis_streams/#component_discarded_events_total](https://vector.dev/docs/reference/configuration/sinks/aws_kinesis_streams/#component_discarded_events_total)

**Reply by jszwedko** [reply]
> Is this where they're defined [src/internal_events/aws_kinesis.rs#L32](https://github.com/vectordotdev/vector/blob/a6b1bed163fa52e1a3fc2189ac24f0f355d752cf/src/internal_events/aws_kinesis.rs#L32)?

That's one of them anyway. There are other shared ones as well in `src/internal_events`.

> I couldn't see anything on the docs for the sink [aws_kinesis_streams/#component_discarded_events_total](https://vector.dev/docs/reference/configuration/sinks/aws_kinesis_streams/#component_discarded_events_total)

Doh, yes, it looks like that is missing from the docs. I'll follow up on that.

**Reply by jszwedko** [reply]
PR to update the docs: https://github.com/vectordotdev/vector/pull/18622
